Validly.Extensions.Validators/Strings/PasswordAttribute.cs

```diff
@@ -60,52 +87,67 @@ public PasswordAttribute(
 	/// A <see cref="ValidationMessage"/> if validation fails; otherwise, <c>null</c>.
 	/// Validation fails if any of the following conditions are met:
 	/// - the value is <c>null</c> or shorter than <see cref="_minLength"/>;
-	/// - <see cref="_requireSpecialChar"/> is <c>true</c> and no special character (non-alphanumeric) is found;
-	/// - <see cref="_requireUpperCase"/> is <c>true</c> and no uppercase letter (A–Z) is present;
-	/// - <see cref="_requireDigit"/> is <c>true</c> and no digit (0–9) is present.
+	/// - the count of special characters is less than <see cref="_numberOfRequiredSpecialCharacters"/>;
+	/// - the count of uppercase letters is less than <see cref="_numberOfRequiredUpperCaseCharacters"/>;
+	/// - the count of digits is less than <see cref="_numberOfRequiredDigitCharacters"/>.
 	/// </returns>
 	[MethodImpl(MethodImplOptions.AggressiveInlining)]
 	public ValidationMessage? IsValid(string? value)
 	{
 		if (value is null || value.Length < _minLength)
 		{
-			return ValidationMessage;
+			return CreateValidationMessage($"Password must be at least {_minLength} characters long.");
 		}
 
-		bool hasSpecialChar = false;
-		bool hasUpperCase = false;
-		bool hasDigit = false;
+		int specialCharCount = 0;
+		int upperCaseCount = 0;
+		int digitCount = 0;
 
 		foreach (char c in value)
```

Comment:
Can you please try to use Span - to iterate over the chars - and benchmark it (compare it)? I'm just curious.

In some builds - when not optimized by the compiler - it will create Enumerator, which will be slow and it can allocate memory IMHO.

When optimized by the compiler, it should be compiled to:
```
        int num = 0;
        while (num < text2.Length)
        {
            char c = text2[num];
            // ...
            num++;
        }
```